agent/src/main/java/io/pyroscope/api/Tags.java

```diff
@@ -0,0 +1,264 @@
+
+        public Snapshot(HashMap<Long, Map<Long, Long>> contexts, HashMap<Long, String> strings) {
+            this.contexts = contexts;
+            this.strings = strings;
+        }
+    }
+    public interface SafeCallable<V> {
+        V call();
+    }
+
+    public static class RefCounted<T> {
+        final Function<T, ValueRef<T>> factory;
+        final ConcurrentHashMap<Long, ValueRef<T>> idToRef = new ConcurrentHashMap<>();
+        final ConcurrentHashMap<T, ValueRef<T>> valueToRef = new ConcurrentHashMap<>();
+
+        private Function<T, ValueRef<T>> myFactory = new Function<T, ValueRef<T>>() {
+            @Override
+            public ValueRef<T> apply(T t) {
+                ValueRef<T> res = factory.apply(t);
+                idToRef.put(res.id, res);
+                return res;
+            }
+        };
+
+        RefCounted(Function<T, ValueRef<T>> factory) {
+            this.factory = factory;
+        }
+
+        ValueRef<T> createRef(T v) {
+            return valueToRef.computeIfAbsent(v, myFactory);
+        }
+
+        public ValueRef<T> getRefById(Long valueId) {
+            return idToRef.get(valueId);
+        }
+
+        public void gc() {
+            Iterator<ValueRef<T>> it = valueToRef.values().iterator();
+            while (it.hasNext()) {
+                ValueRef<T> ref = it.next();
+                if (ref.refCounter.get() == 0) {
+                    it.remove();
+                    idToRef.remove(ref.id);
+                }
+            }
+        }
+    }
+    public static class ValueRef<T> {
+        public final T val;
+        private final AtomicLong refCounter = new AtomicLong(0);
+        public final Long id;
+
+        public ValueRef(T val, Long id) {
+            this.val = val;
+            this.id = id;
+        }
+
+        public long acquire() {
+            return refCounter.incrementAndGet();
+        }
+
+        public long release() {
+            return refCounter.decrementAndGet();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            ValueRef<T> valueRef = (ValueRef<T>) o;
+            return id.equals(valueRef.id);
+        }
+
+        @Override
+        public int hashCode() {
+            return id.hashCode();
+        }
+
+        @Override
+        public String toString() {
+            return "ValueRef{" +
+                "val='" + val + '\'' +
+                ", refCounter=" + refCounter +
+                ", id=" + id +
+                '}';
+        }
+    }
+
+    public static class Scope implements AutoCloseable {
+        final HashMap<Long, Long> prev;
+        final Long prevContextId;
+
+        public Scope(Object... args) { //todo create constructor for 2, 4, 6 args
+            if (args.length %2  != 0) {
+                throw new IllegalArgumentException();
```

Comment:
is it common to provide an error message in this case? (I'm a Java noob)

Reply:
I added error message